File: 백준/Gold/1744. 수 묶기/수 묶기.py
n = int(input())

plus = [] #양수 들 담을거임
minus = [] # 음수들 담을거임

ans = 0
for i in range(n):
    num = int(input())
    if num > 1:
        plus.append(num)
    elif num <= 0:
        minus.append(num)
    else:
        ans += num

plus.sort(reverse=True)
minus.sort()

# 양수 묶기
for i in range(0, len(plus), 2):
    if i+1 >= len(plus):
        ans += plus[i]
    else:
        ans += (plus[i] * plus[i+1])

# 음수 묶기
for i in range(0, len(minus), 2):
    if i+1 >= len(minus):
        ans += minus[i]
    else:
        ans += (minus[i] * minus[i+1])

print(ans)

File: 백준/Gold/1744. 수 묶기/수 묶기.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int n = Integer.parseInt(br.readLine());
		
		List<Integer> minus = new ArrayList<>();
		List<Integer> plus = new ArrayList<>();
		
		for(int i = 0; i < n; i++) {
			int temp = Integer.parseInt(br.readLine());
			if(temp <= 0) {
				minus.add(temp);
			}
			else {
				plus.add(temp);
			}
		}
		Collections.sort(minus);
		Collections.sort(plus, Collections.reverseOrder());
		
		long ret = 0;
		
		for(int i = 0; i < plus.size(); i+=2) {
			if(i +1 >= plus.size()) {
				break;
			}
			//두 수가 같을 때 곱하면 손해임 => 더해
			else if(plus.get(i) == 1 && plus.get(i+1)== 1) {
				ret += plus.get(i) + plus.get(i+1);
				plus.set(i,0);
				plus.set(i+1,0);
			}
			// 곱해 => 근데 둘중에 하나라도 1일 때 곱하면 손해
			else if(plus.get(i) != 1 && plus.get(i+1) != 1){
				ret += plus.get(i) * plus.get(i+1);
				plus.set(i,0);
				plus.set(i+1,0);
			}
		}
		
		for(int i = 0; i < minus.size(); i+=2) {
			if(i +1 >= minus.size()) {
				break;
			}
			
			ret += minus.get(i) * minus.get(i+1);
			minus.set(i,0);
			minus.set(i+1,0);
		}
		for(int i = 0; i < minus.size(); i++) {
			ret += minus.get(i);
		}
		for(int i = 0; i < plus.size(); i++) {
			ret += plus.get(i);
		}
		System.out.println(ret);
	}
}